ciao/arch/architecture.md

```diff
@@ -0,0 +1,423 @@
+is composed of a number of different computing devices referred to as nodes.  It
+can be used to deploy units of computation, known as instances, across its
+nodes.  Once deployed, instances can be monitored and managed.  Two types of
+instances are supported; VM based instances, in which all computation is
+performed inside a hypervisor, and container instances, in which docker
+containers are run directly on the cluster nodes.
+
+Instances are created from workloads and images.  A workload is a set of
+instructions for creating instances of a particular type.  They, for example,
+specify the type of instance to be created, i.e., VM or container, its CPU,
+memory and storage requirements, its backing image and a set of instructions to
+run inside the instance when it is started.  Multiple instances can be created
+from the same workload.  Each instance has its own disk image that contains the
+instance's rootfs.  This disk image is derived from the backing image specified
+in the workload from which it was created.  For VM instances, both the instance
+image and the backing image are stored in a [ceph](http://docs.ceph.com)
+cluster.  For container images, the backing image refers to a docker container
+image supplied by a docker registry.  Both the container backing image and the
+container rootfs image are stored locally on the compute node on which the
+container runs.  Additional data volumes can be created in the ceph cluster and
+attached to both container and VM instances at either creation time or at a
+later date.
+
+Instances, workloads and images are created inside constructs called tenants.
+While the instance is the unit of computation, the tenant is the unit of isolation.
+Instances running in the same tenant are connected together on a tenant private
+network.  Instances that are part of the same tenant can talk to each other but
+those that are part of different tenants, by default, cannot.  This holds true
+even if two instances from different tenants happen to be running on the same
+compute node.  Workloads and images are also, by default, private to the tenants
+in which they were created, although it is possible to create public workloads
+and images that can be used and shared across all tenants.
+
+If the cluster is connected to an external network, instances will be able to
+access services running in that external network.  Although, external client
+applications will not be able to see or talk to instances by default, instances
+can be exposed to the outside world by assigning external IP addresses to them,
+as we shall see below.
+
+Once launched, an instance can be stopped and restarted.  Stopped instances
+typically lose the association with the node on which they previously ran.  By
+stopping and restarting an instance you migrate it from
+one node to another.  Any data stored in data volumes, and for VMs any data
+stored in rootfs volumes, will be migrated intact with the instance.  Live
```

Comment:
Maybe useful to highlight that the network configuration is also preserved across the migration. So when the workload VM or container moves from one node to another, it maintains its IP address and other network configuration. So not only does the DNS identifier not change, its actual IP also does not change,

Reply:
Good point.  I will do.